File: card-lang/src/types/Evaluator.rs

use std::collections::HashMap;

use crate::types::ReusedStructs::Node;
use crate::types::ReusedStructs::NodeType;

pub struct Evaluator {
  constants: HashMap<String, Constant>,
  variables: HashMap<String, Variable>,
  arguments: HashMap<String, Variable>,
  prefixes: HashMap<String, PrefixOperator>,
  operators: HashMap<String, Operator>,
  functions: HashMap<String, Funk>,
}

impl Evaluator {
  pub fn new(
    constants: Vec<Constant>,
    variables: Vec<Variable>,
    prefixes: Vec<PrefixOperator>,
    operators: Vec<Operator>,
    functions: Vec<Funk>,
  ) -> Evaluator {
    let mut evaluator = Evaluator {
      constants: HashMap::new(),
      variables: HashMap::new(),
      arguments: HashMap::new(),
      prefixes: HashMap::new(),
      operators: HashMap::new(),
      functions: HashMap::new(),
    };
    for value in constants {
      evaluator.constants.insert(value.id.clone(), value);
    }
    for value in variables {
      evaluator.variables.insert(value.id.clone(), value);
    }
    for value in prefixes {
      evaluator.prefixes.insert(value.id.clone(), value);
    }
    for value in operators {
      evaluator.operators.insert(value.id.clone(), value);
    }
    for value in functions {
      evaluator.functions.insert(value.id.clone(), value);
    }
    return evaluator;
  }

  fn addConst(&mut self, constant: Constant){
    self.constants.insert(constant.id.clone(), constant);
  }
  fn addVar(&mut self, variable: Variable){
    self.variables.insert(variable.id.clone(), variable);
  }
  fn addPre(&mut self, prefix: PrefixOperator){
    self.prefixes.insert(prefix.id.clone(), prefix);
  }
  fn addOp(&mut self, operator: Operator){
    self.operators.insert(operator.id.clone(), operator);
  }
  fn addFunk(&mut self, funk: Funk){
    self.functions.insert(funk.id.clone(), funk);
  }

  pub fn evaluate(self, parseTree: Vec<Node>) -> String{
    let mut output = "".to_string();
    let mut e: Self = self;
    let mut value: Node;
    for node in parseTree {
      (e, value) = e.parseNode(node);
      output.push_str(Evaluator::nodeToString(value).as_str());
      output.push_str("\n");
    }
    return output;
  }

  fn parseNode(self, node: Node) -> (Self, Node) {
    let mut e: Evaluator = self;
    let n = match node.node_type {
      NodeType::ValueNode=>{
        node
      }
      NodeType::OperatorNode=>{
        let values = node.values.unwrap();
        let op_name = values.get(&"type".to_string()).unwrap();

        let branches = node.branches.unwrap();
        let left = branches.get(&"left".to_string());
        let right: Node;
        (e, right) = self.parseNode(branches.get(&"right".to_string()).unwrap().clone());
        if left.is_none() {
          let prefixRunner = e.prefixes.get(op_name).unwrap().runner;
          prefixRunner.run(right)
        } else {
          let op = e.operators.get(&op_name.to_string()).unwrap().runner;
          let node: Node;
          (e, node) = e.parseNode(left.unwrap().clone());

          return (e, op.run(node, right));
        }
      }
      NodeType::IdentifierNode=>{
        let name = node.values.unwrap().get("value").unwrap().clone();
        let arguments = e.arguments;
        let constants = e.constants;
        let variables = e.variables;
        if arguments.contains_key(&name) {
          return (e, arguments.get(&name).unwrap().value.clone());
        } else if constants.contains_key(&name) {
          return (e, constants.get(&name).unwrap().value.clone());
        } else if variables.contains_key(&name) {
          return (e, variables.get(&name).unwrap().value.clone());
        } else {
          panic!("identifier {} doesn't exist", name);
        }
      }
      NodeType::AssignNode=>{
        let name = node.values.unwrap().get(&"name".to_string()).unwrap().clone();
        let value: Node;
        (e, value) = e.parseNode(
          node.branches.unwrap().get(&"value".to_string()).unwrap().clone()
        );
        e.variables.insert(name.clone(), Variable{ id: name.clone(), value: value.clone() });
        return (e, value)
      }
      NodeType::CallNode=>{
        let mut vals: Vec<Node> = vec![];
        for arg in node.args.unwrap().iter() {
          let result: Node;
          (e, result) = e.parseNode(arg.clone());
          vals.push(result);
        }
        let funk = e.functions.get(
          &node.values.unwrap().get("name").unwrap().clone()
        ).unwrap();
        return (e, funk.runner.run(vals));
      }
      NodeType::FunctionNode=>{
        panic!("Haven't enabled function nodes");
        /*
        let name = node.values.unwrap().get("name").unwrap();

        let names = node.args.unwrap().iter()
        .map(|node| -> String { return *node.values.unwrap().get("name").unwrap() })
        .collect();

        let node = Funk{
          id: *name,
          numArgs: node.args.unwrap().len() as u128,
          runner: Box::new(CreatedFunk {
            evaluator: self,
            arg_names: names,
            node: node
          })
        };

        self.functions.insert(*name, node);
        */
      }
      _=>{
        panic!("invalid node type")
      }
    };
    return (e, n);
  }

  fn nodeToString(self, node: Node) -> String {
    if !matches!(node.node_type, NodeType::ValueNode) {
      panic!("node_type is not ValueNode");
    }
    let values = node.values.unwrap();
    match values["type"].as_str() {
      "boolean" => {
        if values["is_float"].as_str() == "1" {
          return "True".to_string();
        } else {
          return "False".to_string();
        }
      }
      "number" => {
        if values["is_float"].as_str() == "1" {
          return values["value"].as_str().parse::<f64>().unwrap().to_string()
        } else {
          return values["value"].as_str().parse::<i64>().unwrap().to_string()
        }
      }
      "string" => {
        return values.get(&"value".to_string()).unwrap().clone();
      }
      _ => {
        panic!("value typ")
      }
    }
  }
  fn valueToString(self, val: Value) -> String{
    match val {
      Value::BooleanValue(v) => {
        return if v { "True".to_string() } else { "False".to_string() };
      }
      Value::NumberValue(v) => {
        return v.to_string();
      }
      Value::StringValue(v) => {
        return v;
      }
      _ => {
        panic!("invalid value typ")
      }
    }
  }
}

pub enum Value {
  BooleanValue(bool),
  NumberValue(f64),
  StringValue(String),
}

pub struct Constant {
  id: String,
  value: Node
}

pub struct Variable {
  id: String,
  value: Node
}

pub struct PrefixOperator {
  id: String,
  runner: Box<dyn PrefixRunner>
}

pub trait PrefixRunner {
  fn run(&self, a: Node)->Node;
}

pub struct Operator {
  id: String,
  runner: Box<dyn OperatorRunner>
}

pub trait OperatorRunner {
  fn run(&self, a: Node, b: Node)->Node;
}

pub struct Funk {
  id: String,
  numArgs: u128,
  runner: Box<dyn FunkRunner>
}

pub trait FunkRunner {
  fn run(&self, args: Vec<Node> )->Node;
}

struct CreatedFunk {
  evaluator: Evaluator,
  arg_names: Vec<String>,
  node: Node
}

// impl FunkRunner for CreatedFunk {
//   fn run(self, args: Vec<Node>) -> Node {
//     for (i, arg_name) in self.arg_names.iter().enumerate() {
//       self.evaluator.arguments.insert(
//         *arg_name, Variable{ id: *arg_name, value: args[i] }
//       );
//     }
//     let ret = self.evaluator.parseNode(*self.node.branches.unwrap().get("block").unwrap());
//     self.evaluator.arguments.clear();
//     return ret;
//   }
// }

File: card-lang-typescript/src/types/Evaluator.ts
import { Token, LangNode, LangNodeType, ValueNode } from "./ReusedStructs";

export class Evaluator {
  constants: { [key: string]: Constant } = {};

  variables: { [key: string]: Variable } = {};
  arguments: { [key: string]: Variable } = {};

  prefixes:  { [key: string]: PrefixOperator } = {};
  operators: { [key: string]: Operator } = {};
  functions: { [key: string]: Funk } = {};
  constructor(
    constants: Array<Constant>,
    prefixes:  Array<PrefixOperator>,
    operators: Array<Operator>,
    functions: Array<Funk>,
  ){
    for(let value of constants){
      this.constants[value.id] = value;
    }
    for(let value of prefixes){
      this.prefixes[value.id] = value;
    }
    for(let value of operators){
      this.operators[value.id] = value;
    }
    for(let value of functions){
      this.functions[value.id] = value;
    }
  }

  evaluate(parseTree: Array<LangNode>): string{
    return parseTree.map((node)=>{
      return Evaluator.nodeToString(this.parseNode(node))
    }).join("\n")
  }

  private parseNode(node: LangNode): ValueNode {
    switch(node.node_type){
      case LangNodeType.ValueNode: {
        return node
      }
      case LangNodeType.OperatorNode: {
        let op_name = node.values?.type;
        if(op_name == void 0){
          throw new Error("operator node has op type");
        }

        let branches = node.branches;
        if(branches == void 0){
          throw new Error("operator node has no branches");
        }
        let left_branch = branches["left"];
        let right_branch = branches["right"];
        if(right_branch == void 0){
          throw new Error("operator node has no right branch");
        }
        let right = this.parseNode(right_branch);
        if(typeof left_branch == "undefined"){
          return this.prefixes[op_name].runner.run(right)
        } else {
          let left = this.parseNode(left_branch);
          return this.operators[op_name].runner.run(left, right);
        }
      }
      case LangNodeType.IdentifierNode: {
        let name = node.values?.name;
        if(name == void 0){
          throw new Error("identifier node has no name");
        }
        let args = this.arguments;
        let constants = this.constants;
        let variables = this.variables;
        if(name in args){
          return args[name].value;
        }
        if(name in constants){
          return constants[name].value;
        }
        if(name in variables){
          return variables[name].value;
        }
        throw new Error(`Identifier ${name} doesn't exist`);
      }
      case LangNodeType.AssignNode: {
        let name = node.values?.name;
        if(name == void 0){
          throw new Error("identifier node has no name");
        }
        let right = node.branches?.right;
        if(right == void 0){
          throw new Error("identifier node has no values");
        }

        let value = this.parseNode(right);
        this.variables[name] = { id: name, value: value };
        return value;
      }
      case LangNodeType.CallNode: {
        let vals: Array<LangNode> = []
        let args = node.args;
        if(args == void 0){
          throw new Error("no args in callnode");
        }
        let name = node.values?.name;
        if(name == void 0){
          throw new Error("no name in callnode");
        }
        for(let arg of args){
          vals.push(this.parseNode(arg));
        }
        let funk = this.functions[name];
        return funk.runner.run(vals);
      }
      case LangNodeType.FunctionNode: {
        throw new Error("No functions yet");
        /*
        let name = node.values?.name;
        if(name == void 0){
          throw new Error("Function node has no name");
        }

        let names = node.args?.map((arg)=>{
          return arg.values?.name
        })
        if(names == void 0){
          throw new Error("Function node has no args");
        }

        let fn = {
          id: name,
          numArgs: names.length,
          runner: {
            evaluator: this,
            arg_names: names,
            node: node
          }
        };

        this.functions[name] = fn;
        */
      }
      default: {
        throw new Error(`invalid node_type`);
      }
    }
  }

  static nodeToString(node: LangNode): string {
    if(node.node_type != LangNodeType.ValueNode){
      throw new Error("node_type is not ValueNode");
    }
    let values = node.values;
    if(values == void 0){
      throw new Error("node has no values");
    }
    switch(values.type){
      case "boolean": {
        if(values.value == "1"){
          return "True";
        } else {
          return "False";
        }
      }
      case "number": {
        if(values.is_float){
          return Number.parseFloat(values.value).toString()
        } else {
          return Number.parseInt(values.value).toString()
        }
      }
      case "string": {
        return `"${values.value}"`;
      }
      case "function": {
        return "function"
      }
      default: {
        throw new Error("Cannot handle type");
      }
    }
  }
/*
  static valueToString(val: Value) -> String{
    match val {
      Value::BooleanValue(v) => {
        return if v { "True".to_string() } else { "False".to_string() };
      }
      Value::NumberValue(v) => {
        return v.to_string();
      }
      Value::StringValue(v) => {
        return v;
      }
      _ => {
        panic!("invalid value typ")
      }
    }
  }
*/
}

/*
pub enum Value {
  BooleanValue(bool),
  NumberValue(f64),
  StringValue(String),
  FunkValue
}
*/

export type Constant = {
  id: string,
  value: ValueNode
}

type Variable = {
  id: string,
  value: ValueNode
}

export type PrefixOperator = {
  id: string,
  runner: PrefixRunner
}

export interface PrefixRunner {
  run(a: LangNode): ValueNode;
}

export type Operator = {
  id: string,
  runner: OperatorRunner
}

export interface OperatorRunner {
  run(a: LangNode, b: LangNode): ValueNode;
}

export type Funk = {
  id: string,
  numArgs: number,
  runner: FunkRunner
}

export interface FunkRunner {
  run(args: Array<LangNode> ): ValueNode;
}

type CreatedFunk = {
  evaluator: Evaluator,
  arg_names: Array<string>,
  node: LangNode
}

// impl FunkRunner for CreatedFunk {
//   fn run(self, args: Vec<Node>) -> Node {
//     for (i, arg_name) in self.arg_names.iter().enumerate() {
//       self.evaluator.arguments.insert(
//         *arg_name, Variable{ id: *arg_name, value: args[i] }
//       );
//     }
//     let ret = self.evaluator.parseNode(*self.node.branches.unwrap().get("block").unwrap());
//     self.evaluator.arguments.clear();
//     return ret;
//   }
// }